File: apps/ios/AnpusFeather/Game/QA.swift
import CoreGraphics

/*
  The YES / NO question — the port of qa.pde. Answer by tapping a button or by
  moving a hand (touch) into one. Button rects derive from the live world width
  so they stay pinned to the top corners on any screen.
*/
final class QA {
    enum Answer: Equatable { case none, yes, no }

    private(set) var answer: Answer = .none
    private var fillAlpha: CGFloat = 0

    private let btn: CGFloat = 280
    private let margin: CGFloat = 20

    private let store: TextureStore
    private let skeleton: Skeleton
    private var p: [Vec3] { skeleton.points }

    init(store: TextureStore, skeleton: Skeleton) {
        self.store = store
        self.skeleton = skeleton
    }

    private var noRect: CGRect { CGRect(x: margin, y: margin, width: btn, height: btn) }
    private var yesRect: CGRect { CGRect(x: World.w - btn - margin, y: margin, width: btn, height: btn) }

    func answerReset() { answer = .none }

    func ask(_ canvas: Canvas) {
        button(canvas, noRect, "NO", Color4.rgba(255, 0, 0, CGFloat(fillAlpha / 255)))
        button(canvas, yesRect, "YES", Color4.rgba(0, 0, 255, CGFloat(fillAlpha / 255)))
    }

    private func button(_ canvas: Canvas, _ r: CGRect, _ label: String, _ fill: Color4) {
        canvas.fillRect(r.minX, r.minY, r.width, r.height, fill)
        Message.drawCentered(
            canvas, store, label, r.midX, r.midY, fontSize: 60,
            tint: Color4(r: 1, g: 1, b: 1, a: Float(fillAlpha / 255)))
    }

    private func hit(_ x: CGFloat, _ y: CGFloat, _ r: CGRect) -> Bool { r.contains(CGPoint(x: x, y: y)) }

    /// Tap input.
    func enableButtons(_ clicks: [(x: CGFloat, y: CGFloat)]) {
        for c in clicks {
            if hit(c.x, c.y, yesRect) { answer = .yes }
            if hit(c.x, c.y, noRect) { answer = .no }
        }
    }

    /// Hand-position input (skeleton space centered on screen).
    func enableGestures() {
        for hand in [p[4], p[7]] {
            let x = hand.x + World.w / 2
            let y = hand.y + World.h / 2
            if hit(x, y, yesRect) { answer = .yes }
            if hit(x, y, noRect) { answer = .no }
        }
    }

    func fadeIn(_ dt: CGFloat) { fillAlpha = clampf(fillAlpha + dt * FPS, 0, 255) }
    func fadeOut(_ dt: CGFloat) { fillAlpha = clampf(fillAlpha - dt * FPS, 0, 255) }
}

File: apps/unity/Assets/AnpusFeather/Scripts/QA.cs
using UnityEngine;

namespace AnpusFeather
{
    /*
      The YES / NO question — port of qa.pde.
    */
    public sealed class QA
    {
        public enum Answer
        {
            None,
            Yes,
            No
        }

        public Answer CurrentAnswer { get; private set; } = Answer.None;
        float _fillAlpha;

        const float Btn = 280f;
        const float Margin = 20f;

        readonly AssetStore _store;
        readonly Skeleton _skeleton;

        public QA(AssetStore store, Skeleton skeleton)
        {
            _store = store;
            _skeleton = skeleton;
        }

        Rect NoRect => new Rect(Margin, Margin, Btn, Btn);
        Rect YesRect => new Rect(World.W - Btn - Margin, Margin, Btn, Btn);

        public void AnswerReset() => CurrentAnswer = Answer.None;

        public void Ask(DrawCanvas canvas)
        {
            Button(canvas, NoRect, "NO", Color4.Rgba(255f, 0f, 0f, _fillAlpha / 255f));
            Button(canvas, YesRect, "YES", Color4.Rgba(0f, 0f, 255f, _fillAlpha / 255f));
        }

        void Button(DrawCanvas canvas, Rect r, string label, Color4 fill)
        {
            canvas.FillRect(r.xMin, r.yMin, r.width, r.height, fill);
            Message.DrawCentered(canvas, _store, label, r.center.x, r.center.y, 60f,
                new Color4(1f, 1f, 1f, _fillAlpha / 255f));
        }

        static bool Hit(float x, float y, Rect r) => r.Contains(new Vector2(x, y));

        public void EnableButtons((float x, float y)[] clicks)
        {
            foreach (var c in clicks)
            {
                if (Hit(c.x, c.y, YesRect)) CurrentAnswer = Answer.Yes;
                if (Hit(c.x, c.y, NoRect)) CurrentAnswer = Answer.No;
            }
        }

        public void EnableGestures()
        {
            Vec3[] p = _skeleton.Points;
            foreach (int hand in new[] { 4, 7 })
            {
                float x = p[hand].X + World.W / 2f;
                float y = p[hand].Y + World.H / 2f;
                if (Hit(x, y, YesRect)) CurrentAnswer = Answer.Yes;
                if (Hit(x, y, NoRect)) CurrentAnswer = Answer.No;
            }
        }

        public void FadeIn(float dt) => _fillAlpha = MathUtil.Clamp(_fillAlpha + dt * World.Fps, 0f, 255f);
        public void FadeOut(float dt) => _fillAlpha = MathUtil.Clamp(_fillAlpha - dt * World.Fps, 0f, 255f);
    }
}
